modules/csg/csg.cpp

```diff
@@ -278,188 +273,195 @@ void CSGBrush::copy_from(const CSGBrush &p_brush, const Transform3D &p_xform) {
 	_regen_face_aabbs();
 }
 
-// CSGBrushOperation
-
-void CSGBrushOperation::merge_brushes(Operation p_operation, const CSGBrush &p_brush_a, const CSGBrush &p_brush_b, CSGBrush &r_merged_brush, float p_vertex_snap) {
-	// Check for face collisions and add necessary faces.
-	Build2DFaceCollection build2DFaceCollection;
-	for (int i = 0; i < p_brush_a.faces.size(); i++) {
-		for (int j = 0; j < p_brush_b.faces.size(); j++) {
-			if (p_brush_a.faces[i].aabb.intersects_inclusive(p_brush_b.faces[j].aabb)) {
-				update_faces(p_brush_a, i, p_brush_b, j, build2DFaceCollection, p_vertex_snap);
-			}
-		}
+enum {
+	MANIFOLD_PROPERTY_POSITION_X = 0,
+	MANIFOLD_PROPERTY_POSITION_Y,
+	MANIFOLD_PROPERTY_POSITION_Z,
+	MANIFOLD_PROPERTY_NORMAL_X,
+	MANIFOLD_PROPERTY_NORMAL_Y,
+	MANIFOLD_PROPERTY_NORMAL_Z,
+	MANIFOLD_PROPERTY_INVERT,
+	MANIFOLD_PROPERTY_SMOOTH_GROUP,
+	MANIFOLD_PROPERTY_UV_X_0,
+	MANIFOLD_PROPERTY_UV_Y_0,
+	MANIFOLD_MAX
+};
+
+static void _pack_manifold(
+		const CSGBrush *const p_mesh_merge,
+		manifold::Manifold &r_manifold,
+		HashMap<int32_t, Ref<Material>> &p_mesh_materials,
+		Ref<Material> p_default_material,
+		float p_snap) {
+	ERR_FAIL_NULL_MSG(p_mesh_merge, "p_mesh_merge is null");
+
+	HashMap<uint32_t, Vector<CSGBrush::Face>> faces_by_material;
+	for (int face_i = 0; face_i < p_mesh_merge->faces.size(); face_i++) {
+		const CSGBrush::Face &face = p_mesh_merge->faces[face_i];
+		faces_by_material[face.material].push_back(face);
 	}
 
-	// Add faces to MeshMerge.
-	MeshMerge mesh_merge;
-	mesh_merge.vertex_snap = p_vertex_snap;
-
-	for (int i = 0; i < p_brush_a.faces.size(); i++) {
+	manifold::MeshGL64 mesh;
+	mesh.numProp = MANIFOLD_MAX;
+	mesh.runOriginalID.reserve(faces_by_material.size());
+	mesh.runIndex.reserve(faces_by_material.size() + 1);
+	mesh.vertProperties.reserve(p_mesh_merge->faces.size() * 3 * MANIFOLD_MAX);
+
+	// Make a run of triangles for each material.
+	for (const KeyValue<uint32_t, Vector<CSGBrush::Face>> &E : faces_by_material) {
+		const uint32_t material_id = E.key;
+		const Vector<CSGBrush::Face> &faces = E.value;
+		mesh.runIndex.push_back(mesh.triVerts.size());
+
+		// Associate the material with an ID.
+		uint32_t reserved_id = r_manifold.ReserveIDs(1);
```

Comment:
This is a static function? If we do run this once per CSG operation, if I have 1000 CSG brushes at 60 Hz, it will overflow the static uint32_t after just 20 hours of gameplay. If it's once per face, this uint32_t could overflow within minutes.

Is an overflow of the static ID generator expected / normal?

Is there a way to manage the allocation ourselves? I'm just feeling a bit uneasy about using a global static function even if it isn't hammered every frame.

Reply:
are we generating 60 000 * 60 * 60 * 20 boolean operations in a nested tree if operations?

Also this might be an upstream limit until its changed. @elalish Not sure if doubles changed the run id limit

Reply:
I'd like to understand why you need to generate that many IDs. My first instinct is they're probably not being used as intended if you create that many. Their point is generally to be a way to map back to e.g. materials. I doubt you're creating that many unique materials, right?

Reply:
@lyuma As far as I know we only generate one runid per material in a mesh. So I don't know how you are calculating this.